src/ILCompiler.ReadyToRun/src/ObjectWriter/R2RPEBuilder.cs

```diff
@@ -0,0 +1,438 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+
+namespace ILCompiler.PEWriter
+{
+    /// <summary>
+    /// Ready-to-run PE builder combines copying the input MSIL PE executable with managed
+    /// metadata and IL and adding new code and data representing the R2R JITted code and
+    /// additional runtime structures (R2R header and tables).
+    /// </summary>
+    public class R2RPEBuilder : PEBuilder
+    {
+        /// <summary>
+        /// PE reader representing the input MSIL PE file we're copying to the output composite PE file.
+        /// </summary>
+        private PEReader _peReader;
+        
+        /// <summary>
+        /// Complete list of section names includes the sections present in the input MSIL file
+        /// (.text, optionally .rsrc and .reloc) and extra questions injected during the R2R PE
+        /// creation.
+        /// </summary>
+        private ImmutableArray<Section> _sections;
+
+        /// <summary>
+        /// Callback which is called to emit the data for each section.
+        /// </summary>
+        private Func<string, SectionLocation, BlobBuilder> _sectionSerializer;
+
+        /// <summary>
+        /// For each copied section, we store its initial and end RVA in the source PE file
+        /// and the RVA difference between the old and new file. We use this table to relocate
+        /// directory entries in the PE file header.
+        /// </summary>
+        private List<Tuple<int, int, int>> _sectionRvaDeltas;
```

Comment:
If we're using tuples, it would be nice to use the named tuples (also in other places). They're self-documenting.

Reply:
I like the suggestion, will do in a subsequent check-in - I guess it's not fair to ask Simon to polish my clumsy code ;-).

Reply:
I'll let you do this one @trylek since it's not clear to me what the names of the fields should be. I think I'm going to like the named tuple feature.

Reply:
For sure. Please fix the fundamental flaws I made like omission of the copyright notice line - as Michal signed off I assume he doesn't see this as a blocking problem so that I can fix this in a subsequent merge. I'm all for fixing this, I've already mentioned I love his suggestion.